gems/smithy-schema/lib/smithy-schema/document/serializer.rb

```diff
@@ -0,0 +1,283 @@
+          @type_registry = type_registry
+        end
+
+        # Create document data from various input data formats
+        # @param [Object] data Input data can be: Ruby objects, instance of a runtime shape or a
+        #  JSON response with type discriminator.
+        # @return [Data] document data
+        #
+        # @example Ruby objects as input
+        #   # create serializer with an existing type registry
+        #   serializer = Smithy::Schema::Document::Serializer(type_registry)
+        #
+        #   # ruby objects as input
+        #   serializer.create_document("some document")
+        #   # => {"foo" => "bar"}
+        # @example Runtime shape as input
+        #   # create serializer with an existing type registry
+        #   serializer = Smithy::Schema::Document::Serializer(type_registry)
+        #
+        #   # given the following runtime shape
+        #   runtime_shape = some_structure.new.type(some_data)
+        #   # => #<struct SampleService::Types::Structure...>
+        #
+        #   serializer.create_document(runtime_shape)
+        #   # => an instance of Smithy::Schema::Document::Data
+        # @example JSON data
+        #   # create serializer with an existing type registry
+        #   serializer = Smithy::Schema::Document::Serializer(type_registry)
+        #
+        #   # given the following json data
+        #   parsed_json = {
+        #     "__type" => "smithy.ruby.tests#Structure",
+        #     "string" => "hello"
+        #   }
+        #
+        #   document = serializer.create_document(parsed_json)
+        #   # => an instance of Smithy::Schema::Document::Data
+        #   document.discriminator
+        #   # => "smithy.ruby.tests#Structure"
+        def create_document(data)
+          validate_data(data)
+
+          case data
+          when Smithy::Schema::Structure
+            shape = @type_registry.shape_by_type(data.class)
+            Data.new(format_document_data(shape, data), discriminator: shape.id)
+          else
+            if discriminator?(data)
+              shape = @type_registry[data['__type']]
+              Data.new(format_document_data(shape, data, discriminator: true), discriminator: shape.id)
+            else
+              Data.new(serialize_untyped(data))
+            end
+          end
+        end
+
+        # Serializes a document data with optional formatting.
+        # @param [Data] document The document to serialize
+        # @param [Hash] opts serialization options
+        # @option opts [Boolean] :use_timestamp_format Whether to use the
+        #  `timestampFormat` trait or ignore it. The `timestampFormat` trait
+        #   is ignored by default.
+        # @option opts [Boolean] :use_json_name Whether to use `jsonName` trait
+        #   or just member name. The `jsonName` trait is ignored by default.
+        # @return [Hash] Serialized document data
+        def serialize_document(document, opts = {})
```

Comment:
This also feels out of place, perhaps serialized and deserialized methods (or somewhat similar naming like typed/raw) on Document should invoke these paths.